arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,599 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <arch/x86/multiboot.h>
+#include <sys/util.h>
+#include <arch/x86/msr.h>
+#include <kernel_arch_data.h>
+#include <offsets_short.h>
+#include <drivers/interrupt_controller/loapic.h>
+#include <arch/cpu.h>
+
+.section .locore,"ax"
+.code32
```

Comment:
So... the existing x86_64 port has a 32 bit stub that runs as C code and can do stuff like set up early logging, interpret boot parameters (e.g. "am I startup code or a SMP CPU?"), parse a multiboot header (or UEFI if we need to run in a 32 bit UEFI context), set MSRs before mode switch maybe, etc...

The startup code below seems tight and sane, so it's not like it's a problem to do this all in assembly, but it seems like that was a fairly useful feature that got dropped.

Reply:
If you can provide some technical justification for needing 32-bit C here, I'm all ears.  I don't see how any of the things you mention can't be done after we're in long mode.

Reply:
Doesn't this go the other way: what's the technical justifcation for throwing out the C boot stubs given that they're already implemented?  Why is this better than what's already commited?

Reply:
There's no need for this intermediate 32-bit stage or the Byzantine build/link/image process you use to implement it.  You claim this "feature" is "useful" because it parses the multiboot header (by which I assume you mean boot info struct), but that doesn't even work:

```
        /* The multiboot memory map turns out not to be very useful.
         * The basic numbers logged here are only a subset of the true
         * memory map if it has holes or >4G memory, and the full map
         * passed in the second argument tends to live in low memory
         * and get easily clobbered by our own muckery.  If we care
         * about reading memory maps at runtime we probably want to be
         * using BIOS e820 like Linux does.
         */
```
Bizarrely, while you spell out the exact problem (that `xuk` clobbers the memory map before it's read) you seem to imply this is a failing of multiboot, rather than your code.  And your proposed solution to this problem is even worse: fall back into real mode and use the BIOS to read the E820 map?  That not only won't work on many systems (including our targets), it's just doubling down on a flawed design.

Do I really need to justify using my alternative, which, by your own admission, is "sane and tight", and most importantly, works?